### PR TITLE
Fix url in httpGet handler of container lifecycle 

### LINKED ISSUE #56770: hardcode in Url format in httpGet handler of container lifecycle
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
the url is hard code [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/lifecycle/handlers.go#L126) in httpGet handler of container lifecycle.
Thus, I have a pod definition like this:
```
apiVersion: v1
kind: Pod
metadata:
  name: test-pod2
  labels:
    name: test-pod-applied1
spec:
  containers:
  - name: kubernetes-pause
    image: gcr.io/google-containers/nginx
    lifecycle:
      postStart:
        httpGet:
          scheme: HTTPS
          path: /healthz
          port: 8080
```

and the output url looks like this:
```
E1204 14:11:47.181785  116382 handlers.go:128] ----------url: http://172.17.0.3:8080//healthz  
```
The scheme in definition did not take effect, and we got a redundant "/" in path

**What you expected to happen**:
Take advantage of "net/url" package to validate and format the url, instead of hard code.

### PR DESCRIPTION
**Pull request by lichuqiang**
**What this PR does / why we need it**:
Take advantage of "net/url" package to validate and format the url, instead of hard code.

**Which issue(s) this PR fixes**
Fixes #56770

**Special notes for your reviewer**:
/sig node

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/lifecycle/handlers.go [hunk 1]
```diff
@@ -123,8 +124,17 @@ func (hr *HandlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
 			return "", err
 		}
 	}
-	url := fmt.Sprintf("http://%s/%s", net.JoinHostPort(host, strconv.Itoa(port)), handler.HTTPGet.Path)
-	resp, err := hr.httpGetter.Get(url)
+	path := handler.HTTPGet.Path
+	u, err := url.Parse(path)
```
Comment (liggitt) [inline pkg/kubelet/lifecycle/handlers.go]: Don't start by parsing, that changes the meaning of the field from just a path to allowing it to specify a full URL. Construct the URL object as you do below in all cases

File: pkg/kubelet/lifecycle/handlers.go [hunk 2]
```diff
@@ -123,8 +124,17 @@ func (hr *HandlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
...
+	// Something is busted with the path, but it's too late to reject it. Pass it along as is.
+	if err != nil {
+		u = &url.URL{
+			Path: path,
+		}
+	}
+	u.Scheme = string(handler.HTTPGet.Scheme)
```
Comment (liggitt) [inline pkg/kubelet/lifecycle/handlers.go]: The scheme is an enum API value, switch on the values and set to "http" or "https", don't use it directly in the url

File: pkg/kubelet/lifecycle/handlers_test.go [hunk 3]
```diff
@@ -157,7 +158,7 @@ func TestRunHandlerHttp(t *testing.T) {
 	if err != nil {
 		t.Errorf("unexpected error: %v", err)
 	}
-	if fakeHttp.url != "http://foo:8080/bar" {
+	if fakeHttp.url != "HTTP://foo:8080/bar" {
```
Comment (liggitt) [inline pkg/kubelet/lifecycle/handlers_test.go]: Revert this change

File: pkg/kubelet/lifecycle/handlers_test.go [hunk 4]
```diff
@@ -217,7 +219,7 @@ func TestRunHandlerHttpFailure(t *testing.T) {
 	if msg != expectedErr.Error() {
 		t.Errorf("unexpected error message: %q; expected %q", msg, expectedErr)
 	}
-	if fakeHttp.url != "http://foo:8080/bar" {
+	if fakeHttp.url != "HTTP://foo:8080/bar" {
```
Comment (liggitt) [inline pkg/kubelet/lifecycle/handlers_test.go]: Revert this change

File: pkg/kubelet/lifecycle/handlers.go (hunk 2 above)
Comment (liggitt) [reply pkg/kubelet/lifecycle/handlers.go]: Also, consider defaulting to http when the scheme is unspecified

**Comment by lichuqiang** [comment]
@liggitt Updated, PTAL :)

**Comment by lichuqiang** [comment]
@liggitt Looked into funcs in package `net/url`, if we don't do parsing in the beginning, then output func `url.String()` would not work as expect if we specify query in path, say `/echo?msg=poststart` in e2e tests(and the output turn out to be `/echo%3Fmsg=poststart`).
So you would suggest generating url in hardcode way and skip importing package `net/url`?

File: pkg/kubelet/lifecycle/handlers.go [hunk 5]
```diff
@@ -123,8 +124,20 @@ func (hr *HandlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
...
+		Path: handler.HTTPGet.Path,
+		Host: net.JoinHostPort(host, strconv.Itoa(port)),
+	}
+	switch handler.HTTPGet.Scheme {
+	case v1.URISchemeHTTP:
+		u.Scheme = "http"
+	case v1.URISchemeHTTPS:
+		u.Scheme = "https"
+	default:
+		// Defaulting to http when the scheme is unspecified
```
Comment (tallclair) [inline pkg/kubelet/lifecycle/handlers.go]: nit: add the empty string to the "http" case above, and return an error in the default case.

File: pkg/kubelet/lifecycle/handlers.go [hunk 6]
```diff
@@ -123,8 +124,20 @@ func (hr *HandlerRunner) runHTTPHandler(pod *v1.Pod, container *v1.Container, ha
...
+		u.Scheme = "http"
+	}
+	resp, err := hr.httpGetter.Get(u.String())
```
Comment (dixudx) [inline pkg/kubelet/lifecycle/handlers.go]: roundtrip `TLSClientConfig` here to skip verifying insecure `https`?

**Comment by f-ld** [comment]
This PR went to stale state but issue is still real and kinda annoying.
Is someone available to have a look?

**Comment by dixudx** [comment]
/remove-lifecycle stale

@lichuqiang Wanna update this PR?

**Comment by lichuqiang** [comment]
Have not thought of a better way.
At first I meant to do a parsing in the beginning, but seems @liggitt was not so happy with [the behavior](https://github.com/kubernetes/kubernetes/pull/56771#pullrequestreview-88714506)
And if we rely on `url.String()`, as I [described above](https://github.com/kubernetes/kubernetes/pull/56771#issuecomment-357847737), we could run into another problem.
